File: problems/palindrome_list/palindrome_list.py
from typing import Optional


# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        if head is None:
            return True

        stack = []
        curr = head
        while curr is not None:
            stack.append(curr)
            curr = curr.next

        curr = head
        while stack:
            last = stack.pop()
            if curr.val != last.val:
                return False
            curr = curr.next
        return True

File: problems/palindrome_list/PalindromeList.java
package problems.palindrome_list;

import java.util.ArrayDeque;

class ListNode {
    int val;
    ListNode next;

    ListNode() {
    }

    ListNode(int val) {
        this.val = val;
    }
}

public class PalindromeList {
    public boolean isPalindrome(ListNode head) {
        if (head == null)
            return true;
        ArrayDeque<ListNode> stack = new ArrayDeque<>();
        ListNode curr = head;
        while (curr != null) {
            stack.push(curr);
            curr = curr.next;
        }
        curr = head;
        while (!stack.isEmpty()) {
            ListNode last = stack.pop();
            if (curr.val != last.val) {
                return false;
            }
            curr = curr.next;
        }

        return true;

    }
}
